File: ext/lua/examples/intmath.lua
-- ext/lua/examples/intmath.lua —— **四条腿都跑得动的那个子集**（第七个例子家族）
--
-- 期望输出（家族里所有语言、所有后端逐行相同）：15 / 120。
--
-- 这一份是刻意"贫瘠"的：只有整数、函数、调用、语句位置的 if、while、打印一格整数。
-- 理由是量出来的（设计文档 §9）：wasm 那条腿现在**恰好**能接住这些 ——
-- 字符串要线性内存、表达式位置的 if 要 block 带 result、break 要 OIR 有带标签的跳转。
-- 所以这个家族是"图 -> 四个后端"第一次全绿的那一格。

local function sumto(n)
  local acc = 0
  local i = 1
  while i <= n do
    acc = acc + i
    i = i + 1
  end
  return acc
end

local function fact(n)
  if n == 0 then
    return 1
  end
  return n * fact(n - 1)
end

print(sumto(5))
print(fact(5))

File: ext/go/examples/intmath.go
// ext/go/examples/intmath.go —— 与 ext/lua/examples/intmath.lua **同一件事**
//
// 期望输出逐行相同：15 / 120。
// 与那一份一样刻意贫瘠 —— 这是 wasm 那条腿现在能接住的子集（设计文档 §9）。

package main

import "fmt"

func sumto(n int) int {
	acc := 0
	for i := 1; i <= n; i++ {
		acc = acc + i
	}
	return acc
}

func fact(n int) int {
	if n == 0 {
		return 1
	}
	return n * fact(n-1)
}

func main() {
	fmt.Println(sumto(5))
	fmt.Println(fact(5))
}
